src/common/functions-v2/src/scalars/string_multi_args.rs

```diff
@@ -0,0 +1,304 @@
+                name: "concat",
+                args_type: vec![DataType::String; args_type.len()],
+                return_type: DataType::String,
+                property: FunctionProperty::default(),
+            },
+            calc_domain: Box::new(|args_domain, _| {
+                let domain = args_domain[0].as_string().unwrap();
+                Some(Domain::String(StringDomain {
+                    min: domain.min.clone(),
+                    max: None,
+                }))
+            }),
+            eval: Box::new(|args, _generics| {
+                let len = args.iter().find_map(|arg| match arg {
+                    ValueRef::Column(col) => Some(col.len()),
+                    _ => None,
+                });
+                let args = args
+                    .iter()
+                    .map(|arg| arg.try_downcast::<StringType>().unwrap())
+                    .collect::<Vec<_>>();
+
+                let size = len.unwrap_or(1);
+                let mut builder = StringColumnBuilder::with_capacity(size, 0);
+                for idx in 0..size {
+                    for arg in &args {
+                        unsafe { builder.put_slice(arg.index_unchecked(idx)) }
+                    }
+                    builder.commit_row();
+                }
+
+                match len {
+                    Some(_) => Ok(Value::Column(Column::String(builder.build()))),
+                    _ => Ok(Value::Scalar(Scalar::String(builder.build_scalar()))),
+                }
+            }),
+        }))
+    });
+
+    // nullable concat
+    registry.register_function_factory("concat", |_, args_type| {
+        if args_type.is_empty() {
+            return None;
+        }
+        Some(Arc::new(Function {
+            signature: FunctionSignature {
+                name: "concat",
+                args_type: vec![DataType::Nullable(Box::new(DataType::String)); args_type.len()],
+                return_type: DataType::Nullable(Box::new(DataType::String)),
+                property: FunctionProperty::default(),
+            },
+            calc_domain: Box::new(|_, _| None),
+            eval: Box::new(|args, _generics| {
+                type T = NullableType<StringType>;
+                let len = args.iter().find_map(|arg| match arg {
+                    ValueRef::Column(col) => Some(col.len()),
+                    _ => None,
+                });
+                let mut bitmap: Option<Bitmap> = None;
+                let mut inner_args: Vec<ValueRef<StringType>> = Vec::with_capacity(args.len());
+                for arg in args {
+                    let col = arg.try_downcast::<T>().unwrap();
+                    match col {
+                        ValueRef::Scalar(None) => return Ok(Value::Scalar(T::upcast_scalar(None))),
+                        ValueRef::Column(c) => {
+                            bitmap = match bitmap {
+                                Some(m) => Some(common_arrow::arrow::bitmap::and(&m, &c.validity)),
```

Comment:
We can use https://docs.rs/arrow2/latest/arrow2/bitmap/struct.MutableBitmap.html#method.bitand to save an allocation.